sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/WindowedSubscriber.java

```diff
@@ -0,0 +1,929 @@
+        AtomicReferenceFieldUpdater.newUpdater(WindowedSubscriber.class,
+            Subscription.class,
+            "s");
+    private volatile long requested;
+    @SuppressWarnings("rawtypes")
+    private static final AtomicLongFieldUpdater<WindowedSubscriber> REQUESTED =
+        AtomicLongFieldUpdater.newUpdater(WindowedSubscriber.class, "requested");
+    private volatile int wip;
+    @SuppressWarnings("rawtypes")
+    private static final AtomicIntegerFieldUpdater<WindowedSubscriber> WIP =
+        AtomicIntegerFieldUpdater.newUpdater(WindowedSubscriber.class, "wip");
+    private volatile boolean done;
+    private volatile Throwable error;
+    @SuppressWarnings("rawtypes")
+    private static final AtomicReferenceFieldUpdater<WindowedSubscriber, Throwable> ERROR =
+        AtomicReferenceFieldUpdater.newUpdater(WindowedSubscriber.class,
+            Throwable.class,
+            "error");
+
+    /**
+     * Creates the subscriber to split upstream sequence into a series of windows.
+     *
+     * @param loggingContext the context for logging.
+     * @param terminatedMessage the message to include in the error when closing a window after the subscriber termination.
+     * @param nextItemTimout the timeout to trigger window close if the next item for that window does not arrive
+     *     within this period. If {@code null}, this timeout will be disabled.
+     * @param releaser callback to drop the items when there is no active or pending window to consume those.
+     *     If {@code null}, such items will be buffered for future window.
+     */
+    WindowedSubscriber(Map<String, Object> loggingContext, String terminatedMessage, Duration nextItemTimout, Consumer<T> releaser) {
+        this.loggingContext = Objects.requireNonNull(loggingContext, "'loggingContext' cannot be null.");
+        Objects.requireNonNull(terminatedMessage, "'terminatedMessage' cannot be null.");
+        this.terminatedMessage = terminatedMessage + " (Reason: %s)";
+        this.nextItemTimout = nextItemTimout;       // can be null.
+        this.releaser = releaser;     // can be null.
+        this.logger = new ClientLogger(WindowedSubscriber.class, loggingContext);
+    }
+
+    /**
+     * Enqueue a request for a window with up to {@code windowSize} number of items.
+     *
+     * <p>
+     *  TODO (anu) – 1. discuss improving Event Hubs sync receive.
+     *  The WindowedSubscriber can be later moved to azure-core-amqp for both Event Hubs and Service But use.
+     *  For now keeping it in Service Bus to unblock immediate needs.
+     * </p>
+     * <p>
+     *  TODO (anu) – 2. maybe remove 'windowSink' parameter? (discuss with Connie, Liudmila).
+     *  When moving WindowedSubscriber to the azure-core-amqp, do we need to consider passing generic tracer
+     *  interface to trace 'sync receive'?, that enables both Event Hubs and Service Bus tracing, this avoid needing
```

Comment:
we might need to pass tracer around at some point, but we don't do low-level amqp tracing now and you're right, it's best to trace sync receive in the higher levels in the corresponding library.

I believe we pass meter around though, and we should already have in in the azure-core-amqp passed through `ReactorHandlerProvider`. 
In any case, we can add overloads that take tracer/meter overloads later.

Reply:
Thank you for the insight. Now that we discussed, I’ll remove the _todo_